sentinel-core/src/main/java/com/alibaba/csp/sentinel/cluster/TokenService.java

```diff
@@ -44,4 +44,21 @@
      * @return result of the token request
      */
     TokenResult requestParamToken(Long ruleId, int acquireCount, Collection<Object> params);
+
+    /**
+     * Request acquire concurrent tokens from remote token server.
+     *
+     * @param clientAddress the address of the request belong.
+     * @param ruleId ruleId the unique rule ID
+     * @param acquireCount token count to acquire
+     * @return result of the token request
+     */
+    TokenResult requestConcurrentToken(String clientAddress,Long ruleId,int acquireCount);
```

Comment:
Why `clientAddress` is needed here? The token server could get the remote address from Netty context.

And maybe this can be merged in `requestToken`?

Reply:
我们可以通过netty获得地址，TokenService会通过两种方式调用，一个是server调用（本机是server），一种是通过client端发起netty请调用（本机是client）。在我的设计中，如果是server端进行调用，直接从系统传入本机ip，如果client端，这里会传入null，通过netty在server端获取ip地址；两种方式都不会通过网络传输这个地址。之所以会加入这个参数是为了兼容现有的调用形式，因为这个接口会有三个实现，分别是DefaultEmbeddedTokenServer（直接传入本机ip），DefaultClusterTokenClient（传入null），DefaultTokenService（前面两个都会汇入到这里，DefaultClusterTokenClient会从netty获取），如果去掉这个参数DefaultEmbeddedTokenServer，DefaultClusterTokenClient没有影响，但是DefaultTokenService也要去掉这个参数，无法获取地址。没办法沿用以前的接口。

Reply:
我觉得是不能和requestToken合并的，因为我这里需要传入一个token所属client的地址信息，多了一个参数，虽然这个地址信息是不会通过网络传输的。

Reply:
```
@RequestType(ClusterConstants.MSG_TYPE_CONCURRENT_FLOW_ACQUIRE)
public class ConcurrentFlowRequestAcquireProcessor implements RequestProcessor<ConcurrentFlowAcquireRequestData, ConcurrentFlowAcquireResponseData> {
    @Override
    public ClusterResponse processRequest(ChannelHandlerContext ctx, ClusterRequest<ConcurrentFlowAcquireRequestData> request) {
        TokenService tokenService = TokenServiceProvider.getService();
        long flowId = request.getData().getFlowId();
        int count = request.getData().getCount();
        String clientAddress = getRemoteAddress(ctx);
        TokenResult result = tokenService.requestConcurrentToken(clientAddress, flowId, count);
        return toResponse(result, request);
    }
```
这是通过netty获取地址的代码，多了这样一个参数主要还是兼容现在的设计模式。

Reply:
i will make it better in the next pr